### DISCUSSION TITLE
Contact Form File Attachments in Next.js

### DISCUSSION THREAD

**Discussion by DeanParkin** [vercel/next.js#43558] [Help] [answered]
I've spent the last week trying to figure out a method of creating an SMTP contact form that can send a file that a user uploads. I'm just curious if anyone has a preferred method, or if Next offers some function to handle the file upload or if it's potentially an idea the Vercel team would be interested in implementing.

I've not found much in the way that is specific to next on this subject and I'm just wondering if someone could point me in the right dirrection

**Comment by icyJoseph** [comment, marked as answer]
You mean sending `FormData` with content type `Content-Type: multipart/form-data; boundary=something`? If you don't want to upload the image to a CDN that'll exchange it for a URL, and you just want to send it all in one go to an `api` route that'll in turn use nodemailer or whatever to send an email, then you need to use multipart/form-data, and you need to disable the body parser in the `api` route.

Before I post code, or references. Is that what you are trying to do?


## Solution

On a server api/route:

```js
import { Writable } from "stream";
import formidable from "formidable";

const formidableConfig = {
  keepExtensions: true,
  maxFileSize: 10_000_000,
  maxFieldsSize: 10_000_000,
  maxFields: 7,
  allowEmptyFiles: false,
  multiples: false,
};

export const config = {
  api: {
    bodyParser: false,
  },
};

function formidablePromise(req, opts) {
  return new Promise((accept, reject) => {
    const form = formidable(opts);

    form.parse(req, (err, fields, files) => {
      if (err) {
        return reject(err);
      }
      return accept({ fields, files });
    });
  });
}

const fileConsumer = (acc) => {
  const writable = new Writable({
    write: (chunk, _enc, next) => {
      acc.push(chunk);
      next();
    },
  });

  return writable;
};

export default async function handler(req, res) {
  if (req.method !== "POST") return res.status(404).end();

  try {
    const chunks = [];

    const { fields, files } = await formidablePromise(req, {
      ...formidableConfig,
      // consume this, otherwise formidable tries to save the file to disk
      fileWriteStreamHandler: () => fileConsumer(chunks),
    });
    // or the actual field name you used in the front end
    const { careerCV } = files;

    const fileData = Buffer.concat(chunks);

    const filename = careerCV?.originalFilename;

    const attachments = [{ content: fileData, filename }];
    console.log({
      attachments,
      fields,
    });

   // use fileData, fields (that has your form key, value pairs) or attachments as needed
    return res.status(204).end();
  } catch (err) {
    return res.status(500).json({ error: "Internal Server Error" });
  }
}

```

On the client: 

```js
  async function onSubmitForm(data) {
    const formdata = new FormData();

    Object.entries(data).forEach(([key, value]) => {
      if (key === "careerCV") {
        formdata.set(key, value[0]);
      } else {
        formdata.set(key, value);
      }
    });
   // and then use formdata as needed
 }
```

**Reply by DeanParkin** [reply]
yes 100% that's what I need to do

**Reply by DeanParkin** [reply]
https://github.com/DeanParkin/bss-app

This is The repo I'm working on if you want to get a better idea of my setup, the file input is in the CareerForm component and then there is a careersend file in the api folder that is dealing with the upload and send with nodemailer. If you could help me figure this out and learn how to do it I'd be most grateful.

**Reply by icyJoseph** [reply]
Right, I've answered to this a couple of times before, heres one of them: https://github.com/vercel/next.js/discussions/39957

First, to send files over the network, you gotta use FormData, which is only available on browser environments. Adding a form data object to body on a fetch request automatically gives it content-type multipart/form-data. 

On how to do this you can find many articles only, but pretty much you do `const formData = new FormData()`, append each field, and then send it to your API routes with fetch:

```ts
    const res = await fetch('/api/careersend/', {
        method: 'POST',
        body: formData,
    });
```

I see you use axios, but they also have automatic detection for FormData, IIRC.

Second, from the server side. Since this is not a JSON content type, serializing it is not a good option. You gotta tell Next.js to disable the serializer by exporting a config object from the API route, https://nextjs.org/docs/api-routes/request-helpers:

```js
// add this to /api/careersend.js
export const config = {
    api: {
        bodyParser: false,
    },
};
```

Now the body is "unparsed", so we can process it into something meaningful, and collect the file attachment, as well as other fields.

One issue though, FormData doesn't exist in Node! Prior to 18.x.x anyway.  https://nodejs.org/dist/latest-v18.x/docs/api/globals.html#class-formdata

> I assume, but haven't tried, that FormData can be used in Node 18 to consume the request into a FormData object, not sure though.

So what we can do is use a request parser, such as formidable:

```js
import {Writable} from 'stream';

import formidable from 'formidable';

const formidableConfig = {
    keepExtensions: true,
    maxFileSize: 10_000_000,
    maxFieldsSize: 10_000_000,
    maxFields: 7,
    allowEmptyFiles: false,
    multiples: false,
};

function formidablePromise(req, opts) {
    return new Promise((accept, reject) => {
        const form = formidable(opts);

        form.parse(req, (err, fields, files) => {
            if (err) {
                return reject(err);
            }
            return accept({fields, files});
        });
    });
}

const fileConsumer = (acc) => {
    const writable = new Writable({
        write: (chunk, _enc, next) => {
            acc.push(chunk);
            next();
        },
    });

    return writable;
};

export async function handler(req, res) {
    if (req.method !== 'POST') return res.status(404).end();

    try {
        const chunks = [];

        const {fields, files} = await formidablePromise(req, {
            ...formidableConfig,
            // consume this, otherwise formidable tries to save the file to disk
            fileWriteStreamHandler: () => fileConsumer(chunks),
        });

        console.log(fields);
        console.log(files);
        const {file} = files;

        const fileData = Buffer.from(chunks); 
        
        const filename = file?.originalFilename;

        // here you'd pass all the fields and files to the nodemailer transport layer

        const attachments = [{content, filename}];

       const emailRes = await transporter.sendMail({
         from: `${fields.contactName} <${fields.contactEmail}>`,
         // redacted
         subject: `Contact Form Submission from ${fields.contactName}`,
         html: `yada yada`,
         attachments,
       });

        return res.status(204).end();
    } catch (err) {
        return res.status(500).json({error: 'Internal Server Error'});
    }
}

export const config = {
    api: {
        bodyParser: false,
    },
};

export default handler;

```

I reckon for you the missing part is to disable the body parser and read the stream with mutter or whatever tool you have.

**Reply by icyJoseph** [reply]
And maybe I'd move this:

```js
  const transporter = nodemailer.createTransport({
    host: "smtp.gmail.com",
    port: 465,
    secure: true,
    auth: {
      user: process.env.user,
      pass: process.env.pass,
    },
  });
```

Outside of the handler, so that successive requests do not need to instantiate a transporter, and could just re-use a previous instance.

**Reply by DeanParkin** [reply]
I'm so sorry I'm getting really confused with what's going on at each point, I've made another attempt at implementing your code but I'm getting really lost, I've updated my repo with where I got to. I'm so sorry to take up your time but if you could help me I'd be so grateful

**Reply by icyJoseph** [reply]
That's alright, form data can be a bit tricky.

```js
// rest of code...
export default async function handler(req, res) {
  if (req.method !== "POST") return res.status(404).end();

  try {
    const chunks = [];

    const { fields, files } = await formidablePromise(req, {
      ...formidableConfig,
      // consume this, otherwise formidable tries to save the file to disk
      fileWriteStreamHandler: () => fileConsumer(chunks),
    });

    console.log({ fields });
    console.log({ files });
    const { careerCV } = files;

    const fileData = Buffer.from(chunks);

    const filename = careerCV?.originalFilename;

    // here you'd pass all the fields and files to the nodemailer transport layer
    const {
      careerName,
      careerEmail,
      careerMessage,
      careerTelephone,
      careerBadge,
    } = fields;

    const attachments = [{ content: fileData, filename }];
    console.log({
      attachments,
      careerName,
    });

   // removed the transporter part to keep it short
    return res.status(204).end();
  } catch (err) {
    return res.status(500).json({ error: "Internal Server Error" });
  }
}
```

And in your frontend code:

```js
 async function onSubmitForm(data) {
    console.log(data);

    const formdata = new FormData();

    Object.entries(data).forEach(([key, value]) => {
      if (key === "careerCV") {
        formdata.set(key, value[0]);
      } else {
        formdata.set(key, value);
      }
    });
    // formdata.append("careerName", careerName);
    // formdata.append("careerEmail", careerEmail);
    // formdata.append("careerPhone", careerPhone);
    // formdata.append("careerBadge", careerBadge);
    // formdata.append("careerCV", careerCV);
    // formdata.append("careerMessage", careerMessage);

    let config = {
      method: "POST",
      url: `${
        process.env.URL == "http://localhost:3000" ? process.env.URL : ""
      }/api/careersend`,
      data: formdata,
    };

    try {
      const res = await axios(config);
      if (res.status === 200) {
        console.log("Message was sent Successfully");
        messageSent = true;
        //reset();
      }
    } catch (err) {
      console.error(err);
    }
  }
```

### Caveat

If you deploy this to Vercel there's a payload limit:

- https://vercel.com/docs/concepts/limits/fair-use-policy
- https://vercel.com/docs/concepts/limits/overview
- https://vercel.com/docs/concepts/limits/overview#serverless-function-memory

**Reply by icyJoseph** [reply]
@DeanParkin I made an edit to the client side code, when the field is careerCV, we want to take the one file they submitted, and a slight tweak to the server side code to get the `careerCV` file.

**Reply by DeanParkin** [reply]
I'm so sorry dude, I've just tried that code and from what I can tell I'm getting the file to attach and I can see it in the console but when I'm making the http request and I'm getting an error 500

**Reply by DeanParkin** [reply]
ok I have received the email but the file is empty 0mb

**Reply by DeanParkin** [reply]
I think its recognising the fileName but its struggling to accept the file data

**Reply by icyJoseph** [reply]
Ah, now I made a mistake, sorry about that, the attachment must have `{content, filename}` as keys... see here https://nodemailer.com/message/attachments/

So this bit:

```jsx 
 // the key must be `content`
 const attachments = [{ content: fileData, filename }];
```

I'll edit the code above, and that should be it! Do let me know how it goes, hoping for the best!

P.S.: This is why many have moved on to TypeScript. It would've detected this directly.

**Reply by DeanParkin** [reply]
I'm still getting an issue with the transfer, I'm getting this log come through the console:
```
{
  attachments: [
    {
      content: <Buffer 00 00 00 00 00 00 00 00 00 00 00>,
      filename: 'Anchor, Digbeth (Mobile) Lighthouse.pdf'
    }
  ],
  careerName: 'Dean Parkin'
}
```
and that message is coming from this console.log
```
const attachments = [{ content: fileData, filename }];
    console.log({
      attachments,
      careerName,
    });
```
I made a send and an attachment came through but it was only 11 bytes of a 250kb file.

**Reply by icyJoseph** [reply]
Ah 🤦 I often get mixed up here with `.from` or `.concat`, it should be `.concat` !

```js
  const fileData = Buffer.concat(chunks);
```

Last change. Patience, we are almost done!

**Reply by DeanParkin** [reply]
You're a Legend its just worked perfectly, Thank you so much for helping me out bro, I owe you one!!!

**Reply by DeanParkin** [reply]
Damn I got too far ahead, I got it to work in dev mode but I have it running on netlify and its having a wig out

**Reply by icyJoseph** [reply]
Do you have any logs? What's not working out?

**Reply by DeanParkin** [reply]
So I've run it in dev and it works fine but when its in production, I'm getting an error 500 in the browser console saying its got an Axios Error 
when trying to send the form https://bss-security.netlify.app/careers

**Reply by DeanParkin** [reply]
I think axios is having a mare sending it to the api link, but its the correct link so I'm wondering if its something to do with having the correct http content type header for it, I've tried setting it to ```"Content-Type": "multipart/form-data"``` and that hasnt worked

**Reply by icyJoseph** [reply]
I'd start to debug this from the API first. A naive way would be to place logs at every step, and see how far it gets. Do you see any logs about the attachment or anything like that?

I think from the Axios point of view, you could just open the dev tools network tab and when you submit see the headers on the outgoing request. The content type should be set to multipart/form-data. It is of no use to do this manually because there's a boundary-"hash" added to the content type (used to limit each field). Check it out on the dev tools network tab.

**Reply by DeanParkin** [reply]
I'm so sorry, I'm an idiot, I hadn't set the environment variables on the Netlify yet. It all works perfectly, thank you so much 👌

**Reply by OZZlE** [reply]
I have

    <form encType="multipart/form-data; boundry=--">
            <input
                type="file"
                multiple
                required
                accept=".pdf, .docx, .jpg, .png, .xlsx"
                name="what-happened[files]"
              />
    
nextjs v.13.2.4 and latest formidable

I tried this method but I always get an empty `<Buffer />` so no file is attached, looking at the code I don't understand how it deals with multiple files either (it doesn't seem to?)

Running your code server side.

Thank you in advance for help!
